### PR TITLE
nodejs: Add 'build' and 'prune' lifecycle events

### PR DESCRIPTION
**Pull request by djones6**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stacks-overview.md#stack-structure).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

I added `npm build` and `npm prune` steps to the build process for the production image, to provide a point of customization in cases where the project needs to perform build actions (such as running the Angular or Typescript compiler).

The 'prune' step is required to remove any development dependencies, and/or perform any cleanup of artifacts generated during the build, which would otherwise bloat the final image. 

Although there is an `npm prune` command, it does not behave the same way as `install`, `test` etc in that it does not invoke any scripts in package.json (none of `prune`, `preprune`, `postprune` are run).  The only way to run a script called `prune` is to run `npm run prune` (which is what I've gone for).  It is run with `--production` so that the user's script can invoke `npm prune` (possibly in addition to other actions), and this will have the effect of running `npm prune --production`, which removes any development-only dependencies.

Update: I added libssl1.1 and ca-certificates.  This is required in order for the service to connect to others over https (with certificate verification), which I feel is a common enough requirement to be part of the base stack.

### CODE REVIEW

File: incubator/nodejs/image/project/Dockerfile [hunk 1]
```diff
@@ -8,14 +8,29 @@ RUN rm -rf /project/user-app/node_modules && mkdir -p /project/user-app/node_mod
 
 # Install user-app dependencies
 WORKDIR /project/user-app
-RUN npm install --production 
+RUN npm install --production
+
+# Run a build phase. Projects that need to execute build commands can customize the
+# 'build' script in their package.json. The build script should call 'npm install'.
+# If no build is required, just install production dependencies.
```
Comment (neeraj-laad) [inline incubator/nodejs/image/project/Dockerfile:15]: > The build script should call 'npm install'.
> If no build is required, just install production dependencies.

Why is this needed is we are running `npm install --production` ourselves on line 11?

File: incubator/nodejs/image/project/Dockerfile [hunk 2]
```diff
@@ -8,14 +8,29 @@ RUN rm -rf /project/user-app/node_modules && mkdir -p /project/user-app/node_mod
...
+RUN npm run build --if-present
+
+# Uninstall dev dependencies, leaving only production dependencies. Projects can
+# customize the 'prune' script in their package.json.
+# Ideally this would just be 'npm prune', but this command does not have pre/post
+# hooks. Instead, the user's 'prune' script can itself call 'npm prune' in addition
+# to any additional actions required.
+RUN npm run prune --production --if-present
```
Comment (neeraj-laad) [inline incubator/nodejs/image/project/Dockerfile:23]: Again are we delegating specific tasks to developer now? will it work fine if I bring my own app that does not have these scripts defined?

File: incubator/nodejs/image/project/Dockerfile [hunk 3]
```diff
@@ -8,14 +8,29 @@ RUN rm -rf /project/user-app/node_modules && mkdir -p /project/user-app/node_mod
...
 
 # Creating a tar to work around poor copy performance when using buildah 1.9.0
 RUN cd / && tar czf project.tgz project
 
 # Copy the dependencies into a slim Node docker image
 FROM node:12-slim
 
+# Install common dependencies (TLS and CA support)
+RUN apt-get update && apt-get install -y libssl1.1 ca-certificates && apt-get clean
```
Comment (neeraj-laad) [inline incubator/nodejs/image/project/Dockerfile:32]: In the past we've pulled out any apt get updates/upgrades with a view that we want builds to be reproducible. 

Could we install specific versions of these packages?

File: incubator/nodejs/image/project/Dockerfile (hunk 1 above)
Comment (djones6) [reply incubator/nodejs/image/project/Dockerfile:15]: The idea behind these changes was to leave the build for existing apps unchanged - projects that do not define a 'build' and 'prune' script would build as before.

If a build script is created, and the build process requires additional dependencies (defined in `devDependencies`), they would need to be installed at this stage.  The existing `npm install --production` would not install the `devDependencies` that may be necessary for the build phase (for example, a Typescript project would only require the tsc compiler at build time, which emits standard Javascript for runtime).

File: incubator/nodejs/image/project/Dockerfile (hunk 2 above)
Comment (djones6) [reply incubator/nodejs/image/project/Dockerfile:23]: There is no change required for existing apps - the `--if-present` flag allows this command to succeed (do nothing).

If you bring an app that worked in the existing nodejs stack, it will continue to work in this version.  If you bring an app that requires a build step (such as running the `tsc` compiler), you would now be able to use it with this stack by doing something like this in your `package.json`:
```
    "build": "npm install && tsc",
    "prune": "npm prune"
```
This works because `npm run build` is run without the `--production` flag, and so calling `npm install` will install the `devDependencies` defined in the package.
When `npm run prune` is run, the `--production` flag sets environment which propagates to the `npm prune` command.  This removes any dependencies that would not be installed by `npm install --production`.

File: incubator/nodejs/image/project/Dockerfile (hunk 3 above)
Comment (ianpartridge) [reply incubator/nodejs/image/project/Dockerfile:32]: `FROM node:12-slim` is not reproducible anyway though, is it? That image will be updated regularly by upstream.

Comment (djones6) [reply incubator/nodejs/image/project/Dockerfile:32]: Similar to what Ian just commented, I'm not sure it would make sense to install more specific versions than these.  This would match (for example) the packages that would be published in the `node:12` (full fat) image.  We'd want to pick up security fixes to all packages automatically, and as I understand it, the way to do this is to regularly republish the image.

If there's a desire to consume those updates in a more selective / scheduled way, I think that's a separate issue and probably needs some tooling around it (like an intermediate slim image that a stack is based on, with its own release / versioning process).

(as an aside, we're also installing these packages in the nodejs-express stack: https://github.com/appsody/stacks/blob/master/incubator/nodejs-express/image/project/Dockerfile#L29 )

File: incubator/nodejs/README.md [hunk 4]
```diff
@@ -40,6 +40,20 @@ Templates are used to create your local project and start your development. When
 
     **NOTE:** Currently the `appsody deploy` command only works for deploying web applications.
 
+## Customizing the build
+
+Simple Node projects do not require a 'build' process: they are executed directly from source.  However, some technologies such as Typescipt require commands to be invoked to prepare the project ready to be run.
```
Comment (BethGriggs) [inline incubator/nodejs/README.md:45]: nit: Typescrip -> TypeScript
